pallets/stateful-storage/src/types.rs

```diff
@@ -450,26 +702,31 @@ impl<T: Config> ItemizedOperations<T> for ItemizedPage<T> {
 		let mut count = 0u16;
 		let mut items = BTreeMap::new();
 		let mut offset = 0;
+		let page_size = self.data.len();
+
 		while offset < self.data.len() {
 			ensure!(
-				offset + ItemHeader::max_encoded_len() <= self.data.len(),
-				PageError::ErrorParsing("wrong header size")
+				offset + ItemHeader::max_encoded_len() <= page_size,
+				PageError::ErrorParsing("incomplete item header")
 			);
 			let header = <ItemHeader>::decode(&mut &self.data[offset..])
-				.map_err(|_| PageError::ErrorParsing("decoding header"))?;
-			let item_total_length = ItemHeader::max_encoded_len() + header.payload_len as usize;
+				.map_err(|_| PageError::ErrorParsing("decoding item header"))?;
+			let item_total_length = ItemHeader::max_encoded_len() + header.payload_len() as usize;
```

Comment:
I don't think this is going to work. Previously ItemHeader had a deterministic max size but now there is difference in size for V1 and V2 and to be able to do correct max check you need to first discern the type and then get the max for that type.